### ISSUE TITLE
Nmin/Vorfruchtwert aus Gemüse wird nicht auf HF2 (Gemüse) gerechnet bzw. ist Verhalten kurios

### ISSUE THREAD

**Issue by simonkriegnerschramml** [w3geo/bml-fast#121]
Hallo Arnold,

folgende Situation:
Ich habe eine HF1 Gemüsekultur "Kraut (Industrie, früh)" und eine HF2 Gemüsekultur "Zwiebel (Sommer, trocken)".

In die Nmin Zelle der HF2 wird korrekt der Nmin Defaultwert von 95 kg N eingetragen. Nur passiert damit nichts. Dieser wird nicht in die N-Bilanzberechnungen der HF2 eingerechnet. Erst wenn ich den Wert editiere, reagiert auch die Berechnung. Kurios ist aber auch wenn ich die ursprünglichen 95 kg wieder händisch eintrage, dann verschwindet der Wert wieder aus der Bilanz. Komischerweise wird in einem anderen Beispiel alles korrekt umgesetzt (siehe Beispielfläche "N-Saldo VF-HF1 u. HF1-HF2")

Siehe Beispielfläche bei der es nicht funktioniert "Gemüse auf Gemüse". Ich weiß beim besten Willen nicht wo da die Ursache begraben sein könnte.

[fast-export-2024-11-19.txt](https://github.com/user-attachments/files/17812380/fast-export-2024-11-19.txt)

LG
Simon

**Comment by arnold-pichler** [comment]
Da habe ich jetzt auch etwas gebraucht, aber der Unterschied zwischen den Fällen ist die Bilanz. Beim "Gemüse auf Gemüse" hast du KEINE DÜNGUNG und damit keinen Bilanzüberschuss - damit fällt das quasi in eine andere Zeile der Regel-Matrix.

Ich fürchte fast, der Fehler liegt jetzt in der Korrektur eines anderen Fehlers der letzten Version ( #118 ). Muss jetzt vorsichtig sein, dass ich durch die Korrektur nicht wieder einen anderen Fehler neu reinbringe. Deshalb sind die Testfälle so wichtig und die automatischen Tests.

**Comment by arnold-pichler** [comment]
OK, da ist noch was nicht korrekt umgesetzt aus der Matrix. Sehe ich das richtig, hier ist der Fall:

![image](https://github.com/user-attachments/assets/5359de11-30a2-463a-9834-0aaa5c7cbfbf)

D.h. Bilanzüberschuss ist <= 20, Gemüse auf Gemüse, kein manueller NMIN, dann müßten die 95 eigentlich bei der HF2 als Vorfruchtwert aus der HF1 angeführt sein, korrekt?

![image](https://github.com/user-attachments/assets/196f21d6-bade-4573-949f-36a62199c843)

**Comment by simonkriegnerschramml** [comment]
Bitte aufpassen. Der Fehler liegt nur in der Tatsache, dass ein Nmin, welcher gleich dem ausgelesenen Wert ist, nicht angerechnet wird. irgendwie scheint mir, dass hier ein Vergleich, was denn nun strenger sei, gemacht wird (95=95). Das brauchen wir hier aber nicht, da der manuelle Nmin immer den ausgelesenen Wert einer Gemüse Vorkultur schlägt. Ein Vergleich muss erst dann erfolgen, falls es einen N-Saldo geben sollte. Dann muss der Wert, der in der Zelle Nmin steht, dem N-Saldo gegenübergestellt werden.
Wichtig ist aber, wenn gar kein N-Saldo vorhanden sein sollte, dass dieser Nmin Wert trotzdem angewandt wird. dieser ist ja dann größer als ein N-Saldo von 0 kg.

LG
Simon

**Comment by simonkriegnerschramml** [comment]
> OK, da ist noch was nicht korrekt umgesetzt aus der Matrix. Sehe ich das richtig, hier ist der Fall:
> 
> ![image](https://private-user-images.githubusercontent.com/36012148/388476073-5359de11-30a2-463a-9834-0aaa5c7cbfbf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxODkxNzAsIm5iZiI6MTczMjE4ODg3MCwicGF0aCI6Ii8zNjAxMjE0OC8zODg0NzYwNzMtNTM1OWRlMTEtMzBhMi00NjNhLTk4MzQtMGFhYTVjN2NiZmJmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIxVDExMzQzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThhODQwNTZmYTU0NzQ4MGZhNDNiZWU2YmMyMGFlZWI1ZjhiM2MwMjhkYjY1NTkwZmMwZjdiNzU0MTBiODExODAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.5pCh76UsIzhXhjSZqiF6W_yyiIShDFDNze753ww5BiA)
> 
> D.h. Bilanzüberschuss ist <= 20, Gemüse auf Gemüse, kein manueller NMIN, dann müßten die 95 eigentlich bei der HF2 als Vorfruchtwert aus der HF1 angeführt sein, korrekt?

Ja, das ist korrekt. Aber wir bezeichnen diesen Vorfruchtwert bei Gemüsekulturen immer als Nmin. Egal ob dieser nun aus einer Messung stammt oder eben aus der Spalte "VFW|Nmin ...." ausgelesen wird. Ich glaube das ist sogar schon so umgesetzt. Erst wenn die Folgefrucht eine Nicht Gemüse Kultur sein sollte, dann heißt dieser Wert aus der Spalte "VFW|Nmin ..." "Vorfruchtwert.

LG
Simon
> 
> ![image](https://private-user-images.githubusercontent.com/36012148/388475831-196f21d6-bade-4573-949f-36a62199c843.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIxODkxNzAsIm5iZiI6MTczMjE4ODg3MCwicGF0aCI6Ii8zNjAxMjE0OC8zODg0NzU4MzEtMTk2ZjIxZDYtYmFkZS00NTczLTk0OWYtMzZhNjIxOTljODQzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIxVDExMzQzMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4MzlkOWQ3OGZkMThkYTg1ZmU1MjhiOGRkYjc3M2Y4MWE5Y2Y3OWRhNzYyNzExNGM4OGVlMWQ2NzJmMWY4YTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.YT0vdzORIT4k2PODgJv91zEyfOKk54fFko4QA6kifOs)

**Comment by arnold-pichler** [comment]
> Aber wir bezeichnen diesen Vorfruchtwert bei Gemüsekulturen immer als Nmin

OK - in der Excel ist es als Vorfruchtwert bezeichnet, auch für Gemüse -> Gemüse:

![image](https://github.com/user-attachments/assets/aed11945-b6a1-4c58-826f-8cbb0947c15b)

**Comment by simonkriegnerschramml** [comment]
Sorry Kommando retour. Wir lassen das wording so. Das verwirrt den user


Disclaimer und Datenschutz der Landwirtschaftskammer Oberösterreich
Diese Nachricht ist vertraulich. Im Falle einer Fehlleitung, verständigen Sie bitte den Absender und löschen Sie diese Nachricht und alle Anhänge. Sollten Sie nicht der beabsichtigte Adressat sein, ist jede Offenlegung, Weiterleitung oder sonstige Verwendung dieser Information nicht gestattet. Soweit gesetzlich zulässig, schließt der Absender jede Haftung für Schäden aus Übertragungsfehlern, Viren, fremden Einfluss, Verzögerungen und dergleichen aus. Nähere Informationen zum Datenschutz finden Sie unter:  www.ooe.lko.at/datenschutz<http://www.ooe.lko.at/datenschutz>
Am 21.11.2024 13:40 schrieb Arnold Pichler ***@***.***>:

Aber wir bezeichnen diesen Vorfruchtwert bei Gemüsekulturen immer als Nmin

OK - in der Excel ist es als Vorfruchtwert bezeichnet, auch für Gemüse -> Gemüse:

image.png (view on web)<https://github.com/user-attachments/assets/aed11945-b6a1-4c58-826f-8cbb0947c15b>

**Comment by simonkriegnerschramml** [comment]
> > Dieser Wert ist unabhängig von einer etwaigen Düngung.
> 
> Ja, eh. Aber in dem Fall entscheidet die Düngung darüber, welchen Fall aus der Matrix ich behandle (wg. der Bilanz). Und der Fall für Bilanz <= 20 ist einfach noch fehlerhaft von mir umgesetzt bzw. durch die Korrektur eines anderen Fehlers kam es hier zu einer Regression.
> 
Ich verstehe nicht warum hier die Düngung entscheiden sollte. Die Düngung ist ja nur ein Teil des anrechenbaren Stickstoffs und wenn keine vorhanden sein sollte, dann ist dieser Teil eben 0. In der Matrix gibt es auch gar keine Bedingung, welche auf die Düngung abzielt.

> > irgendwie scheint mir, dass hier ein Vergleich, was denn nun strenger sei, gemacht wird (95=95)
> 
> Nein. Aber es wird dadurch entschieden, ob ein manueller NMin Wert existtiert, denn wenn der Eingegebene Wert nicht der Vorgabe entspricht, ist es ein manueller Wert.

**Comment by simonkriegnerschramml** [comment]
> > Aber wir bezeichnen diesen Vorfruchtwert bei Gemüsekulturen immer als Nmin
> 
> OK - in der Excel ist es als Vorfruchtwert bezeichnet, auch für Gemüse -> Gemüse:

Kommando retour. Wir lassen die Bezeichnung Vorfruchtwert stehen. Das führt nur zu Verwirrung beim Testen und beim user schließlich auch. So sehen wir auf den ersten Blick, welcher Wert nun in die Berechnung einfließt.

LG
Simon

> ![image](https://private-user-images.githubusercontent.com/36012148/388513927-aed11945-b6a1-4c58-826f-8cbb0947c15b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIyMDA1MzYsIm5iZiI6MTczMjIwMDIzNiwicGF0aCI6Ii8zNjAxMjE0OC8zODg1MTM5MjctYWVkMTE5NDUtYjZhMS00YzU4LTgyNmYtOGNiYjA5NDdjMTViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIxVDE0NDM1NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNmYmVlNjc0MWQ3NDhiN2RjMWVhMTUwYjljMzJlOTE4OWMxZTM5ZTk3ODE0M2M3ZGViMGJmNDkyYTFjODJmYjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2kVH4RcJWtFelWMl3LuWdaQWdQDs9M-uPW-Ppf6FgHI)

**Comment by arnold-pichler** [comment]
> Ich verstehe nicht warum hier die Düngung entscheiden sollte. Die Düngung ist ja nur ein Teil des anrechenbaren Stickstoffs und wenn keine vorhanden sein sollte, dann ist dieser Teil eben 0. In der Matrix gibt es auch gar keine Bedingung, welche auf die Düngung abzielt.

Ist die nicht im Bilanzüberschuss?? Ich habe das so interpretiert. Wenn das nicht stimmt, haben wir hier ohnehin einen gröberen Fehler drin.

**Comment by simonkriegnerschramml** [comment]
> > Ich verstehe nicht warum hier die Düngung entscheiden sollte. Die Düngung ist ja nur ein Teil des anrechenbaren Stickstoffs und wenn keine vorhanden sein sollte, dann ist dieser Teil eben 0. In der Matrix gibt es auch gar keine Bedingung, welche auf die Düngung abzielt.
> 
> Ist die nicht im Bilanzüberschuss?? Ich habe das so interpretiert. Wenn das nicht stimmt, haben wir hier ohnehin einen gröberen Fehler drin.

Das Problem ist, dass der Bilanzüberschuss aus mehreren Zugängen und einem einzigen Abgang besteht. Die Düngung ist nur ein einziger Part der Zugänge und keine Bedingung für eine Bilanzierung. Als Beispiel kann ich folgendes anbieten. Es gibt eine Obergrenze des Bilanzüberschusses von 100 kg, daraus resultieren in Feuchtgebieten maximal 60 kg und in Trockengebieten maximal 80 kg N-Saldo ja ha. Wenn wir bspw. noch einen Vorfruchtwert der Vorfrucht aus dem vorangegangenen Jahr generieren, kann der anrechenbare Stickstoff noch höher ausfallen. Dabei haben wir noch nicht einmal gedüngt. Jetzt nehmen wir HF1 wie Soja, Ackerbohnen oder Futterzwischenfrüchte deren Düngeobergrenze je ha ja nicht gerade hoch ausfällt. Dann haben wir einen Bilanzüberschuss generiert ohne auch nur ein kg an Stickstoff aktiv gedüngt zu haben.

folgendes Beispiel:

Vorfrucht: Ackerbohne - Vorfruchtwert 20 kg
N-Saldo: BAS16=60 kg
Futter-Zwischenfrucht 1-60% Leg. - Düngeobergrenze außerhalb Nitratrisikogebiet: 40 kg

N-Bilanz der Futterzwischenfrucht: 20+ 60 - 40 = 40 kg Bilanzüberschuss
40>20 kg , daher kommt es zur N-Saldierung: 40x0,6=24 kg

LG
Simon

**Comment by arnold-pichler** [comment]
> keine Bedingung für eine Bilanzierung

Das ist klar, aber ich meinte nur, dass die Düngung da AUCH drin ist und in den beiden Beispielen spielte sie eine Rolle.

Die Bilanz ist doch letztlich das da, korrekt?

![image](https://github.com/user-attachments/assets/166130d8-7613-41c1-8253-bd7458444fa9)

**Comment by simonkriegnerschramml** [comment]
> > keine Bedingung für eine Bilanzierung
> 
> Das ist klar, aber ich meinte nur, dass die Düngung da AUCH drin ist und in den beiden Beispielen spielte sie eine Rolle.
> 
> Die Bilanz ist doch letztlich das da, korrekt?
> 
> ![image](https://private-user-images.githubusercontent.com/36012148/388838653-166130d8-7613-41c1-8253-bd7458444fa9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIyNjEwOTEsIm5iZiI6MTczMjI2MDc5MSwicGF0aCI6Ii8zNjAxMjE0OC8zODg4Mzg2NTMtMTY2MTMwZDgtNzYxMy00MWMxLTgyNTMtYmQ3NDU4NDQ0ZmE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIyVDA3MzMxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYxNjU4ZTMyZGNkYjhhYTg3OGFmZmM1OTA3MmJlNWVjNGU2N2U2MTQwN2Q3Yzc4MDBiNDE0NjVjZTYzZGIxOWMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.WXW5J2PWuv9tYslahvXZ1lxKZeg2JWVZwWXWJkZWRac)

Genau. Es muss aber nicht jeder einzelne Part existent sein, damit man bilanzieren kann. Wenn ein Part nicht vorhanden sein sollte, dann ist dieser schlichtweg 0.

LG
Simon

**Comment by arnold-pichler** [comment]
> dann ist dieser schlichtweg 0.

Ja, sowieso. Das ist schon klar. Dann hat schon alles seine Richtigkeit, aber es ist eben ein Fehler drin in der Umsetzung. Da bin ich dran.

**Comment by arnold-pichler** [comment]
Ich geh gerade nochmal alle Fälle durch. Nur zur Sicherheit, ich habe das vermutlich schon in einem anderen Issue gefragt - hier (gelb markiert) ist der `VFW | Nmin selbes Jahr` der HF2 gemeint, korrekt? Nicht der der HF1 (der wird für den Vorfruchtwert gebraucht).

![image](https://github.com/user-attachments/assets/24e36405-b1f5-4d4e-8a72-b4eb98e548fd)

**Comment by simonkriegnerschramml** [comment]
> Ich geh gerade nochmal alle Fälle durch. Nur zur Sicherheit, ich habe das vermutlich schon in einem anderen Issue gefragt - hier (gelb markiert) ist der `VFW | Nmin selbes Jahr` der HF2 gemeint, korrekt? Nicht der der HF1 (der wird für den Vorfruchtwert gebraucht).
> 
> ![image](https://private-user-images.githubusercontent.com/36012148/388881227-24e36405-b1f5-4d4e-8a72-b4eb98e548fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIyNjg1NTksIm5iZiI6MTczMjI2ODI1OSwicGF0aCI6Ii8zNjAxMjE0OC8zODg4ODEyMjctMjRlMzY0MDUtYjFmNS00ZDRlLThhNzItYjRlYjk4ZTU0OGZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTIyVDA5MzczOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWY4Y2MwZjY5OTJhZGFkNDhhMmEwNTc0ZGZhZTdmYTczMDc4NDAzNzNmNzI1MmY0MDk4YTYzODNlMDc0MGQ2MjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tK_wxx4Z7zfqhDHnuGsYRadGL0gRw-nvGiQmDy2ZlIk)

Nein hier muss der VFW der HF 1 für die Anzeige bei der HF 2 herangezogen werden. Der VFW der Vorfrucht aus dem Vorjahr ist für die HF 1 relevant.

LG
Simon

**Comment by arnold-pichler** [comment]
> Nein hier muss der VFW der HF 1 für die Anzeige bei der HF 2 herangezogen werden. Der VFW der Vorfrucht aus dem Vorjahr ist für die HF 1 relevant.

Von der Vorfrucht rede ich hier gar nicht. Das ist glaube ich sowieso eines unserer Probleme, dass der Begriff "Vorfrucht" unterschiedliche Bedeutungen in unserem Kontext hat, aber egal jetzt.

Ich gehe gerade nur nochmal das Sheet "N_HF1 auf HF2" durch und darin alle Fälle. Hier wird ja systematisiert, was für die HF2 angerechnet wird. Da habe ich in zwei Spalten den `Wert VFW | Nmin selbes Jahr` , wobei es bei "Vorfruchtwert aus der HF1" logischerweise den für die HF1 meint (roter 2er im Screenshot). Aber was ist beim roten 1er gemeint - der von HF1 oder HF2? Ich war der Meinung, dass hier mit dem von HF2 verglichen werden muss?

![image](https://github.com/user-attachments/assets/fc7f4e4e-2bf1-4ab4-95f5-e5017f0ebc29)

**Comment by arnold-pichler** [comment]
Ich habe die gesamte Matrix "N_HF 1 auf HF 2..." neu überprüft, neu implementiert - Bitte durchtesten und am besten dabei gleich die Testfälle machen!

**Comment by simonkriegnerschramml** [comment]
Ich denke, das Issue kann geschlossen werden. Nach umfangreichen Testungen, habe ich vorerst keine Fehler mehr entdecken können.

LG
Simon
